common/util/trees.h

```diff
@@ -0,0 +1,856 @@
+#define VERIBLE_COMMON_UTIL_TREES_H_
+
+#include <algorithm>
+#include <cstddef>
+#include <functional>
+#include <iosfwd>  // IWYU pragma: keep
+#include <iterator>
+#include <numeric>
+#include <set>
+#include <utility>
+#include <vector>
+
+#include "common/util/logging.h"
+#include "common/util/spacer.h"
+#include "common/util/type_traits.h"
+
+namespace verible {
+
+namespace trees_internal {
+
+// TODO(mglb): move (Unavailable)FeatureTraits to a more generic header, or
+// replace them with something from standard library.
+
+// Type used as a placeholder for unavailable feature in type traits.
+struct UnavailableFeatureTraits {
+  static inline constexpr bool available = false;
+};
+
+// Type used as a base for available feature in type traits.
+struct FeatureTraits {
+  static inline constexpr bool available = true;
+};
+
+// TreeNodeTraits implementation details:
+
+// Alias to Node::subnodes_type if it exists.
+// Intended for use with detected_or_t.
+template <typename Node,  //
+          typename Type_ = typename Node::subnodes_type>
+using TreeNodeSubnodesType = Type_;
+
+// Defined when `Node` contains `Children()` method which returns reference to
+// a STL-like container with nodes.
+// TODO(mglb): Add support for children containers storing Node pointers.
+template <typename Node,  //
+          typename ChildrenType_ = decltype(std::declval<Node>().Children()),
+          typename =
+              std::void_t<decltype(*std::declval<Node>().Children().begin()),
+                          decltype(std::declval<Node>().Children().end())>>
+struct TreeNodeChildrenTraits : FeatureTraits {
+  // Container type which should be used for storing arrays of detached
+  // child nodes. It must be move-assignable to, and move-constructible from,
+  // a value returned by `Children()` method.
+  using container_type = detected_or_t<std::remove_reference_t<ChildrenType_>,
+                                       TreeNodeSubnodesType, Node>;
+};
+
+// Defined when `Node` contains `Value()` method.
+template <typename Node,  //
+          typename Value_ = decltype(std::declval<Node>().Value())>
+struct TreeNodeValueTraits : FeatureTraits {
+  // Type of data returned by `Value()` method.
+  using type = std::remove_reference_t<Value_>;
+  // Aliases provided for convenience
+  using reference = std::add_lvalue_reference_t<type>;
+  using const_reference = std::add_lvalue_reference_t<const type>;
+};
+
+// Defined when `Node` contains `Parent()` method which dereferences to a type
+// fulfilling NodeTraits concept.
+template <typename Node,  //
+          typename Parent_ = decltype(*std::declval<Node>().Parent()),
+          typename = std::void_t<TreeNodeChildrenTraits<Parent_>>>
+struct TreeNodeParentTraits : FeatureTraits {};
+
+// BirthRank implementation details:
+
+// BirthRank implementation supportin any container
+template <class T>
+inline static size_t BirthRank(const T& node, std::input_iterator_tag) {
```

Comment:
I'd remove `inline` and `static` because these are already templates in a public header.

Reply:
Fixed.